olp-cpp-sdk-authentication/src/AuthenticationClient.cpp

```diff
@@ -769,6 +901,74 @@ client::CancellationToken AuthenticationClient::Impl::SignOut(
   });
 }
 
+client::CancellationToken AuthenticationClient::Impl::IntrospectApp(
+    std::string access_token, IntrospectAppCallback callback) {
+  using ResponseType =
+      client::ApiResponse<IntrospectAppResult, client::ApiError>;
+
+  auto introspect_app_task =
+      [=](client::CancellationContext context) -> ResponseType {
+    if (!settings_.network_request_handler) {
+      client::ApiError result({static_cast<int>(http::ErrorCode::IO_ERROR),
+                               "Cannot introspect app while offline"});
+      return std::move(result);
+    }
+
+    client::AuthenticationSettings auth_settings;
+    auth_settings.provider = [&access_token]() { return access_token; };
+
+    client::OlpClientSettings settings;
+    settings.network_request_handler = settings_.network_request_handler;
+    settings.task_scheduler = settings_.task_scheduler;
+    settings.proxy_settings = settings_.network_proxy_settings;
+    settings.authentication_settings = auth_settings;
+
+    client::OlpClient client;
+    client.SetBaseUrl(settings_.token_endpoint_url);
+    client.SetSettings(settings);
+
+    std::multimap<std::string, std::string> query_params;
+    std::multimap<std::string, std::string> header_params;
+    std::multimap<std::string, std::string> forms_params;
+    std::string content_type;
+    auto http_result = client.CallApi(kIntrospectAppEndpoint, "GET",
+                                      query_params, header_params, forms_params,
+                                      nullptr, content_type, context);
+
+    rapidjson::Document document;
+    rapidjson::IStreamWrapper stream(http_result.response);
+    document.ParseStream(stream);
+    if (http_result.status != http::HttpStatusCode::OK) {
+      // HttpResult response can be error message or valid json with it.
+      std::string msg = http_result.response.str();
+      if (!document.HasParseError() && document.HasMember(Constants::MESSAGE)) {
+        msg = document[Constants::MESSAGE].GetString();
+      }
+      client::ApiError error({http_result.status, msg});
+      return std::move(error);
+    }
+
+    IntrospectAppResult response_result = GetIntrospectAppResult(document);
+    return std::move(response_result);
+  };
+
+  // wrap_callback needed to convert client::ApiError into AuthenticationError.
+  auto wrap_callback = [callback](ResponseType response) {
+    if (!response.IsSuccessful()) {
+      auto error = response.GetError();
+      AuthenticationError auth_error{error.GetHttpStatusCode(),
+                                     error.GetMessage()};
+      callback(std::move(auth_error));
```

Comment:
Instead, probably you can merge and save the extra move:

```suggestion
      callback({error.GetHttpStatusCode(), error.GetMessage()});
```